invoker/core/src/main/java/com/google/cloud/functions/invoker/TypedFunctionExecutor.java

```diff
@@ -0,0 +1,195 @@
+    return executor;
+  }
+
+  /**
+   * Returns the {@code T} of a concrete class that implements {@link BackgroundFunction
+   * BackgroundFunction<T>}. Returns an empty {@link Optional} if {@code T} can't be determined.
+   */
+  static Optional<Type> handlerTypeArgument(Class<? extends TypedFunction<?, ?>> functionClass) {
+    return Arrays.stream(functionClass.getMethods())
+        .filter(method -> method.getName().equals(APPLY_METHOD) && method.getParameterCount() == 1)
+        .map(method -> method.getGenericParameterTypes()[0])
+        .findFirst();
+  }
+
+  /** Executes the user's method, can handle all HTTP type methods. */
+  @Override
+  public void service(HttpServletRequest req, HttpServletResponse res) {
+    HttpRequestImpl reqImpl = new HttpRequestImpl(req);
+    HttpResponseImpl respImpl = new HttpResponseImpl(res);
+    ClassLoader oldContextClassLoader = Thread.currentThread().getContextClassLoader();
+
+    try {
+      Thread.currentThread().setContextClassLoader(func.getClass().getClassLoader());
+      handleRequest(reqImpl, respImpl);
+    } finally {
+      Thread.currentThread().setContextClassLoader(oldContextClassLoader);
+      flushResponse(respImpl);
+    }
+  }
+
+  private void handleRequest(HttpRequest req, HttpResponse resp) {
+    WireFormat format = configuration.getFormat();
+
+    Object reqObj = null;
+    try {
+      reqObj = format.deserialize(req, argType);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to parse request for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Object respObj = null;
+    try {
+      respObj = this.func.apply(reqObj);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to execute " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    try {
+      format.serialize(respObj, resp);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to serialize response for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+  }
+
+  private void flushResponse(HttpResponseImpl respImpl) {
+    try {
+      // We can't use HttpServletResponse.flushBuffer() because we wrap the
+      // PrintWriter
+      // returned by HttpServletResponse in our own BufferedWriter to match our API.
+      // So we have to flush whichever of getWriter() or getOutputStream() works.
+      try {
+        respImpl.getOutputStream().flush();
+      } catch (IllegalStateException e) {
+        respImpl.getWriter().flush();
+      }
+    } catch (IOException e) {
+      // Too bad, can't flush.
+    }
+  }
+
+  private static class ConfigurationImpl implements TypedFunction.Configuration {
+    private WireFormat format = null;
+
+    WireFormat getFormat() {
+      if (format == null) {
```

Comment:
sorry just saw this, reason for checking format null twice?

Reply:
Yep! This is a somewhat common pattern for thread-safe lazy instantiation of a singleton. The first check is done without holding the lock, there is a race condition here but because we only ever set this value from null to non-null (and never back) it is safe. The second check with the lock is slower but safer and ensures that there can be no other threads attempting to set the format field at the same time.

Reply:
I think the [holder idiom](https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom) for this would be a bit more straightforward.